python/ray/_private/worker.py

```diff
@@ -2097,13 +2097,13 @@ def connect(
             " and will be removed in the future."
         )
 
-    # Start the import thread
+    # Setup import thread, but defer the start up of 
+    # import thread until job_config is initialized.
+    # (python/ray/_raylet.pyx maybe_initialize_job_config)
     if mode not in (RESTORE_WORKER_MODE, SPILL_WORKER_MODE):
         worker.import_thread = import_thread.ImportThread(
```

Comment:
Are there any cases where other parts of Ray will call things like `import_thread.join()` during this window where the import thread isn't started?

I also notice that `self.threads_stopped.is_set()` [is checked _after_](https://github.com/ray-project/ray/blob/3122b9631abd398f6acc609c363dc9e5b91b2aaf/python/ray/_private/import_thread.py#L67) `self._do_importing()` is called. So in the case where `threads_stopped` is set before the import thread is started, then the import thread will call `_do_importing` once before shutting down. Could cause a crash. I'm not sure how this is handled or even if it's likely.

Reply:
> Are there any cases where other parts of Ray will call things like import_thread.join() during this window where the import thread isn't started?

this should be fine since the `import_thread.join()` will check if import_thread has already started.

Reply:
> I also notice that self.threads_stopped.is_set() [is checked after](https://github.com/ray-project/ray/blob/3122b9631abd398f6acc609c363dc9e5b91b2aaf/python/ray/_private/import_thread.py#L67) self._do_importing() is called. So in the case where threads_stopped is set before the import thread is started, then the import thread will call _do_importing once before shutting down. Could cause a crash. I'm not sure how this is handled or even if it's likely.

good catch.

Reply:
> > Are there any cases where other parts of Ray will call things like import_thread.join() during this window where the import thread isn't started?
> 
> this should be fine since the `import_thread.join()` will check if import_thread has already started.

Yep. I was thinking more of ordering issues -- any thread which calls `import_thread.join` will expect to continue after import_thread has finished. But that assumption could not be held. Not sure how impactful this is; hence why I asked.